packages/algob/src/lib/compile.ts

```diff
@@ -85,3 +92,87 @@ export class CompileOp {
     fs.writeFileSync(filename, content);
   }
 }
+
+export class PyCompileOp {
+  compileOp: CompileOp;
+
+  constructor (compileOp: CompileOp) {
+    this.compileOp = compileOp;
+  }
+
+  /**
+   * Description : returns compiled teal code from pyTeal file
+   * @param filename : name of the PyTeal code in `/assets` directory.
+   *                   Examples : [ gold.py, asa.py]
+   *                   MUST have .py extension
+   * @param force    : if true it will force recompilation even if the cache is up to date.
+   */
+  async ensureCompiled (filename: string, force: boolean): Promise<PyASCCache> {
+    if (!filename.endsWith(pyExt)) {
+      throw new Error(`filename "${filename}" must end with "${pyExt}"`);
+    }
+
+    const content = this.compilePyTeal(filename);
+    const [teal, thash] = [content, murmurhash.v3(content)];
+
+    const a = await this.readArtifact(filename);
+    if (!force && a !== undefined && a.srcHash === thash) {
+      // '\x1b[33m%s\x1b[0m' for yellow color warning
+      console.warn('\x1b[33m%s\x1b[0m', `smart-contract source "${filename}" didn't change, skipping.`);
+      return a;
+    }
+    console.log("compiling", filename);
+    const compiledTeal = await this.compileOp.compile(filename, teal, thash);
+    const pyCompiled: PyASCCache = {
+      filename: "",
+      timestamp: 0,
+      compiled: "",
+      compiledHash: "",
+      srcHash: 0,
+      tealCode: content,
+      toBytes: new Uint8Array(1)
+    };
+    Object.assign(pyCompiled, compiledTeal);
+
+    const cacheFilename = path.join(CACHE_DIR, filename + ".yaml");
+    this.compileOp.writeFile(cacheFilename, YAML.stringify(pyCompiled));
+    return pyCompiled;
+  }
+
+  async readArtifact (filename: string): Promise<PyASCCache | undefined> {
+    await assertDir(CACHE_DIR);
+    try {
+      const p = path.join(CACHE_DIR, filename + ".yaml");
+      return YAML.parse(await fs.promises.readFile(p, 'utf8')) as PyASCCache;
+    } catch (e) {
+      if (e?.errno === -2) { return undefined; } // handling a not existing file
+      throw e;
+    }
+  }
+
+  /**
+   * Description: Runs a subprocess to execute python script
+   * @param filename : python filename in assets folder
+   */
+  private runPythonScript (filename: string): SpawnSyncReturns<string> {
+    // used spawnSync instead of spawn, as it is synchronous
+    return spawnSync('python3', [
+      path.join(ASSETS_DIR, filename)],
+    { encoding: 'utf8' }
+    );
+  }
+
+  /**
+   * Description: returns TEAL code using pyTeal compiler
+   * @param filename : python filename in assets folder
+   */
+  compilePyTeal (filename: string): string {
+    const subprocess: SpawnSyncReturns<string> = this.runPythonScript(filename);
+
+    if (subprocess.stderr) {
+      console.error(subprocess.stderr);
+      throw new Error(subprocess.stderr);
```

Comment:
Please use Algob errors. Maybe you will need to add a new error to `src/internal/core/errors-list.ts`

Reply:
ok